pkg/pusher/pusher.go

```diff
@@ -129,35 +125,44 @@ func (s *Service) chunksWorker(warmupTime time.Duration, tracer *tracing.Tracer)
 	}
 
 	push := func(op *Op) {
+		defer func() {
+			wg.Done()
+			<-sem
+		}()
+
 		s.metrics.TotalToPush.Inc()
 		ctx, logger := ctxLogger()
 		startTime := time.Now()
-		wg.Add(1)
-		go func() {
-			defer func() {
-				wg.Done()
-				<-s.sem
-			}()
-			if err := s.pushChunk(ctx, op.Chunk, logger, op.Direct); err != nil {
-				// warning: ugly flow control
-				// if errc is set it means we are in a direct push,
-				// we therefore communicate the error into the channel
-				// otherwise we assume this is a buffered upload and
-				// therefore we repeat().
-				if op.Err != nil {
-					op.Err <- err
-				}
-				repeat()
-				s.metrics.TotalErrors.Inc()
-				s.metrics.ErrorTime.Observe(time.Since(startTime).Seconds())
-				loggerV1.Debug("cannot push chunk", "chunk_address", op.Chunk.Address(), "error", err)
-				return
+
+		if err := s.valid(op.Chunk); err != nil {
+			logger.Warning("stamp with is no longer valid, skipping syncing for chunk", "batch_id", hex.EncodeToString(op.Chunk.Stamp().BatchID()), "chunk_address", op.Chunk.Address(), "error", err)
+			ctx, cancel := context.WithTimeout(ctx, chunkStoreTimeout)
+			defer cancel()
+			if err = s.storer.Set(ctx, storage.ModeSetSync, op.Chunk.Address()); err != nil {
```

Comment:
This only needs to be done if it is deferred push as this updates the pushIndex and other localstore entries. These are not present for direct push chunks.

For direct push we need to return error on the op.Err channel like below.